Add TraverseBreadthFirst and TraverseDepthFirst methods

fsateler opened the pull request:
These help in traversing trees

atifaziz commented on MoreLinq/Traverse.cs:
```diff
@@ -0,0 +1,97 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
```
Blank line after for clarity, please.

atifaziz commented on MoreLinq/Traverse.cs:
```diff
@@ -0,0 +1,97 @@
+        /// Generates a sequence by traversing an element according to the <paramref name="nextSteps"/> generator function. 
+        /// The returned elements are traversed Breadth First.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
```
Type parameter doc missing.

atifaziz commented on MoreLinq/Traverse.cs:
```diff
@@ -0,0 +1,97 @@
+        /// Generates a sequence by traversing an element according to the <paramref name="nextSteps"/> generator function. 
+        /// The returned elements are traversed Depth First.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
```
Type parameter doc missing here too.

atifaziz commented on MoreLinq/Traverse.cs:
```diff
@@ -0,0 +1,97 @@
+        /// </remarks>
+        public static IEnumerable<T> TraverseBreadthFirst<T>(T seed, Func<T, IEnumerable<T>> nextSteps)
+        {
+            if (nextSteps == null) {
```
Collapse trivial argument checks into one line like this, making it consistent with the rest of the project style:
```c#
            if (nextSteps == null) throw new ArgumentNullException("nextSteps");
```

atifaziz commented on MoreLinq/Traverse.cs:
```diff
@@ -0,0 +1,97 @@
+            queue.Enqueue(seed);
+
+            while (queue.Count != 0) {
+                var cur = queue.Dequeue();
```
`cur` is unusual. Consider `curr` or `current` instead for something more conventional.

atifaziz commented on MoreLinq/Traverse.cs:
```diff
@@ -0,0 +1,97 @@
+        /// </remarks>
+        public static IEnumerable<T> TraverseDepthFirst<T>(T seed, Func<T, IEnumerable<T>> nextSteps)
+        {
+            if (nextSteps == null) {
```
Collapse trivial argument checks into one line as shown earlier.

atifaziz commented on MoreLinq/Traverse.cs:
```diff
@@ -0,0 +1,97 @@
+            while (queue.Count != 0) {
+                var cur = queue.Dequeue();
+                yield return cur;
+                foreach (var child in nextSteps(cur)) {
```
If you do go with BSD-style brace layout then you can skip it here for single-statement `for` loop.

atifaziz commented on MoreLinq/Traverse.cs:
```diff
@@ -0,0 +1,97 @@
+    {
+        /// <summary>
+        /// Generates a sequence by traversing an element according to the <paramref name="nextSteps"/> generator function. 
+        /// The returned elements are traversed Breadth First.
```
> The returned elements are traversed Breadth First.

This reads confusing because it says how the user of the result will traverse the result whereas the returned elements are the results of a breadth-first traversal. How about this?

> The returned elements are the result of breadth-first traversal.

atifaziz commented on MoreLinq/Traverse.cs:
```diff
@@ -0,0 +1,97 @@
+            var queue = new Queue<T>();
+            queue.Enqueue(seed);
+
+            while (queue.Count != 0) {
```
Hate to bring this up, but if it's not too painful, please use BSD-style brace layout to be consistent with the rest of the project style but I won't hold you to it as the implementation is short & straightforward. :wink: Plus I don't enjoy playing the style-cop.

atifaziz commented on MoreLinq/Traverse.cs:
```diff
@@ -0,0 +1,97 @@
+    public partial class MoreEnumerable
+    {
+        /// <summary>
+        /// Generates a sequence by traversing an element according to the <paramref name="nextSteps"/> generator function. 
```
There seems to be general vocabulary/wording/model confusion between `TraverseBreadthFirst` and `TraverseDepthFirst` being generators or tree traversal algorithms. I think you need to settle on one and then have consistent wording and naming. I'd say that since the operators speak of traversing, the tree model is what gets evoked first (at least did for me and I'm imagining it'll be the same for others). But then `seed` and `nextSteps` evoke a generator model. Within the implementation, the loop `child` variable (albeit being a local variable) evokes the tree once again. The doc summaries talk in terms of both, which made it confusing to read and review.

fsateler commented:
I have pushed a new version addressing your comments, and rebased on top of current master to avoid the merge conflict.

I don't know how to mark the reviews as fixed...

atifaziz commented on MoreLinq/Traverse.cs:
```diff
@@ -0,0 +1,92 @@
+    public partial class MoreEnumerable
+    {
+        /// <summary>
+        /// Traverses a tree starting at the <paramref name="root"/> according to the <paramref name="childrenSelector"/> traversal function.
```
Summaries should avoid referencing parameters via `<paramref>` elements because in a [table of summaries](https://morelinq.github.io/1.x/ref/api/html/T_MoreLinq_MoreEnumerable.htm), a reader may not have enough context so using plain language is best.

Consider instead:

> Traverses a tree in breadth-first fashion, starting at a root node and using a user-defined function to get the children at each node of the tree.

atifaziz commented on README.md:
```diff
@@ -73,6 +73,8 @@ Operator             | Summary
 `ThenBy`             | Performs a subsequent ordering of elements in a sequence in a particular direction (ascending, descending) according to a key. This method has 2 overloads.
 `ToDataTable`        | Appends elements in the sequence as rows of a given object with a set of lambda expressions specifying which members (property or field) of each element in the sequence will supply the column values. This method has 4 overloads.
 `ToDelimitedString`  | Creates a delimited string from a sequence of values. The delimiter used depends on the current culture of the executing thread. This method has 30 overloads.
+`TraverseBreadthFirst` | Traverses a tree in a Breadth First manner
```
“Breadth First” is not a proper noun so should read “breadth-first“ instead. I would suggest just using the text from the XML doc summary.

atifaziz commented on README.md:
```diff
@@ -73,6 +73,8 @@ Operator             | Summary
 `ToDataTable`        | Appends elements in the sequence as rows of a given object with a set of lambda expressions specifying which members (property or field) of each element in the sequence will supply the column values. This method has 4 overloads.
 `ToDelimitedString`  | Creates a delimited string from a sequence of values. The delimiter used depends on the current culture of the executing thread. This method has 30 overloads.
+`TraverseBreadthFirst` | Traverses a tree in a Breadth First manner
+`TraverseDepthFirst` | Traverses a tree in a Breadth First manner
```
“Depth First” is not a proper noun so should read “depth-first“ instead. I would suggest just using the text from the XML doc summary.

atifaziz commented on MoreLinq/Traverse.cs:
```diff
@@ -0,0 +1,92 @@
+        /// <returns>A sequence containing the traversed values</returns>
+        /// <remarks>
+        /// This function defers traversal until needed and streams the results.
+        /// The progression function is not checked for loops. If you need the resulting sequence to be finite
```
What's the _progression function_?

> If you need the resulting sequence to be finite you must ensure that loops are not produced by `childrenSelector`

Avoid using. Consider re-phrasing like this:

> If the resulting sequence needs to be finite then it is the responsibility of `childrenSelector` to ensure that loops are not produced.

fsateler commented:
@atifaziz I have addressed the new comments. I seem to have missed the Test comments, and I can't see it either here or on my mail. Could you repost that comment?

atifaziz commented on MoreLinq.Test/TraverseTest.cs:
```diff
@@ -0,0 +1,128 @@
+            res.AssertSequenceEqual(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
+        }
+
+        private class Tree<T>
```
`Tree<T>` seems to be defined in way to benefit from collection & object initialization syntax later on in the actual test code but I find that if you define it simply as it should be:

```c#
        class Tree<T>
        {
            public T Value { get; }
            public IEnumerable<Tree<T>> Children { get; }

            public Tree(T value, IEnumerable<Tree<T>> children)
            {
                Value = value;
                Children = children;
            }
        }
```

and define a helper as follows:

```c#
        static class Tree
        {
            public static Tree<T> New<T>(T value, params Tree<T>[] children) =>
                new Tree<T>(value, children);
        }
```

then the test code reads simpler & more concise & the tree still looks visually like one:

```c#
        [Test]
        public void TraverseBreadthFirstTraversesBreadthFirst()
        {
            var tree =
                Tree.New(1,
                    Tree.New(2,
                        Tree.New(3)),
                    Tree.New(5,
                        Tree.New(6)));

            var res = MoreEnumerable.TraverseBreadthFirst(tree, t => t.Children).Select(t => t.Value);
            res.AssertSequenceEqual(1, 2, 5, 3, 6);
        }

        [Test]
        public void TraverseDepthFirstTraversesDepthFirst()
        {
            var tree =
                Tree.New(1,
                    Tree.New(2,
                        Tree.New(3)),
                    Tree.New(5,
                        Tree.New(6)));

            var res = MoreEnumerable.TraverseDepthFirst(tree, t => t.Children).Select(t => t.Value);
            res.AssertSequenceEqual(1, 2, 3, 5, 6);
        }
```

fsateler replied:
OK, thanks. I have implemented this now in a new commit.

atifaziz commented:
> I seem to have missed the Test comments, and I can't see it either here or on my mail. Could you repost that comment?

That's odd but okay, I've [re-posted](https://github.com/morelinq/MoreLINQ/pull/192/files#r79341838). Let me know if you got it this time.

fsateler commented:
@atifaziz sorry, that happens when one uses a different editor :unamused: 

Force pushed a fixed version.

atifaziz commented:
Closed by ae619b070fe654ea95187b930b9401bd89835fb8

@fsateler Thanks for contributing & working through the reviews.

atifaziz commented:
This is now release as part of [2.0 beta 5](https://www.nuget.org/packages/morelinq/2.0.0-beta05).

atifaziz commented:
Docs published with morelinq/morelinq.github.io@fcbf33be40d5054a9921e49da36b3409981928cf refresh:

- [`TraverseBreadthFirst`](https://morelinq.github.io/2.0/ref/api/html/M_MoreLinq_MoreEnumerable_TraverseBreadthFirst__1.htm)
- [`TraverseDepthFirst`](https://morelinq.github.io/2.0/ref/api/html/M_MoreLinq_MoreEnumerable_TraverseDepthFirst__1.htm)
